### ISSUE TITLE
[FR] Per-unit teamcolor

### ISSUE THREAD

**Issue by sprunk** [beyond-all-reason/spring#949]
Spot the hero unit:
![image](https://github.com/beyond-all-reason/spring/assets/2573076/61f41f84-8f8c-454c-b5bc-c7a6dc4d7b44)

Having special units denoted by a custom teamcolor is a classic convention. Not just heroes either, for example a game could denote buffs or elemental affinity or whatever this way.

The request is to either expose `Spring.SetUnitTeamColor(unitID, r, g, b, a)` or, if this is already possible via shaders, to add a widget with such shader (and a similar WG/GG API) to basecontent for easy inclusion.

**Comment by Krogoth100** [comment]
I've already done it for myself, my code is incompatible with BAR, but its a widget solution.
Just use one of the userDefined for render object SSBO:
    uint composite {u8 drawFlag, u8 unused1, u16 id},
    float maxHealth,
    float health,
    vec4 drawPos,
    vec4 speed,
    vec4[] userDefined
Look widgets that work with engine SSBO, set either full color, or index (spend 1 slot instead of 3 and address some uniform array).

**Comment by lhog** [comment]
I think we can piggyback on teamcolor abstraction and assign an individual unit to a virtual team (color). This way no changes are required in the current pipeline, but this would also mean the amount of virtual team colors is going to be limited (say another 256 teams).
How does it sound?

**Comment by sprunk** [comment]
So basically a palette? Like:
```lua
Spring.SetCustomPaletteTeamcolor(13, {1,0,0,1}) -- 13th palette color is now red
Spring.SetUnitCustomPaletteTeamcolor(unitID, 13) -- unit is now red (sticks across capture / team change too)
Spring.SetPaletteColor(999, {1,1,1,1}) -- error, palette doesn't have 999 slots
Spring.SetUnitCustomPaletteTeamcolor(unitID, nil) -- unit is now back to whatever is its "real" color
```
Sounds fine.

---

With two proposed solutions maybe I should write more about my goals here because I don't know which of the two would be better.

My primary goal is to make it easy for modders and/or small gamedevs to be able to make a unit look different, programmatically, since often they can't count on differentiating them "properly" either via model or writing a shader themselves. This is from experience (both personal from back when I was a noob modder, but also looking at other successful platforms like wc3/sc2 modding). I think the goal is fairly important, despite it often resulting in units that look pretty low quality or even outright ugly, because that is how people generally start out, can't expect everybody to be a shader master (or generally to do much other than tweak a config file that defines a colour) from day 1.

There are two specific facets to that primary goal:
 * one is to be able to adjust teamcolor, which is this ticket (in the engine repo).
 * the other is to tint the whole unit, for which I made a *gameside* ticket in ZK ZeroK-RTS/Zero-K#4988 which could then be put into basecontent if it turned out well (ignore the scaling part for now, because that has other consequences than just rendering).

In this vein, putting the teamcolor change feature engine-side makes sense, because for various reasons that's what's most newbie-friendly, and suggests perhaps I should have made the tinting request engine-side as well. In particular when Krogoth100 says his shader is "not compatible with BAR" it's somewhat worrisome because that is where most modders are going to come from, while an engine feature is almost necessarily compatible with every game.

But: my second goal is to keep things "ideologically clean", i.e. keep things on the abstraction level where they belong. I did not make an engine request for tinting because I assume it should be doable gameside easily by someone who knows what they're doing via some sort of `gl.Color` or similar calls, and I didn't think teamcolor would be. If Krogoth100 says he made it work then perhaps it would point to the correct solution being some basecontent wupget, as the engine already exposes sufficient API.

But: my third goal is to keep things modern and performant. `gl.Color` is literally GL1 so maybe it should be discouraged, with better GL4 solutions around? But that brings us back to the "newbies cant write this" issue. Note that I've been keeping the "doable via `gl.Color`" assumption, but as a rendering layman I don't actually know this and would have to do trial and error. So perhaps it's fine if the engine exposes a seemingly redundant API if this redundancy achieves performance and/or ease of use.

So ultimately, I am still unsure (for either teamcolor or tinting) where this belongs. I'd somewhat lean towards engine-side but both approaches sound good.

**Comment by lhog** [comment]
It's possible to do with custom widget like @Krogoth100 did it, but it's not the best solution here:
1) The engine rendering (for whoever not using CUS or CUS-GL4) won't be able to render these units with expected custom teamcolor.
2) In order to scale the effect every shader drawing this particular unit needs to be updated to leverage this custom SSBO attribute.
In general SSBO attributes are made for tasks like that, to enable setting arbitrary values per unit/feature and make them available on GPU for rendering, but here I feel the task might be worth the more standardized solution, like the API you proposed for example.

**Comment by Krogoth100** [comment]
It _may_ be good to do it on engine side. The reason why I myself do it with Lua however, because I don't know how much flexibility I will end up using. For example, I may end up using 2 team colors to define a unit, remember W40K Dawn of War color schematics. Putting stuff in engine is rigid, lua is flexible.

**Comment by lhog** [comment]
Yes, makes sense to do in Lua in some cases, like the above

**Comment by Krogoth100** [comment]
Maybe, for simplicity, remake engine SSBO's teamColors -> paletteColors.
(Remake means just rename that word)
Indeed add Spring.SetPaletteColor (simplified Spring.SetTeamColor).
Custom indices will start from the last team index.

**Comment by lhog** [comment]
NB: I checked the code and adding extra fake teams is non-trivial. it's rather easier to extract color interface from Team / BaseTeam class.

**Comment by schmied** [comment]
Hi guys! As a toy project I am currently writing a color generator with arbitrary team and team-member count. Goal is to evenly distribute hue in perceptional space (OKLab). Language is C++, a small class, but should be portable. No further requirements, only some calculations. So may I ask if this code has chance to get integrated? Please, can you hint me to the code location where colors are generated on your side? Thanks!

Fun fact: Yesterday I watched a reply where literally an enemy Tik destroyed an AFU and because of the very same color of friend and enemy the one guy didn't notice and got fucked. Look it up, it was a high profile game from Sunday or so, and the incident happened in the lower left corner.

**Comment by p2004a** [comment]
@schmied I don't think this issue is for that. This is just a feature request to be able to set **per-unit** team color in engine. What you probably wanted comment on is https://github.com/beyond-all-reason/Beyond-All-Reason/issues/2439 and best to share how those would look like.

**Comment by schmied** [comment]
@p2004a Thanks! I searched the wrong repo.

**Comment by schmied** [comment]
FYI

![Image](https://github.com/user-attachments/assets/066a7575-ad64-4301-9e75-b3f2591ce359)
